compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -75,186 +75,490 @@ std::optional<uint32_t> getNumberIterations(scf::ForOp loop) {
-  auto constant = rewriter.create<arith::ConstantOp>(
-      rewriter.getUnknownLoc(), rewriter.getIndexAttr(bdId.value()));
-  AMDAIE::BdIdOp bdIdOp = rewriter.create<AMDAIE::BdIdOp>(
-      rewriter.getUnknownLoc(), tileOp, constant.getResult());
-  return bdIdOp;
-};
+  /// Release a given BdId op.
+  LogicalResult releaseBdId(AMDAIE::BdIdOp bdIdOp) {
+    auto tileOp =
+        dyn_cast_if_present<AMDAIE::TileOp>(bdIdOp.getTile().getDefiningOp());
+    if (!tileOp) {
+      return bdIdOp.emitOpError()
+             << "doesn't operate on a `amdaie.tile` operation";
+    }
 
-LogicalResult releaseBdId(
-    AMDAIE::BdIdOp bdIdOp,
-    DenseMap<Value, ChannelBdIdGenerator> &shimTileToGeneratorMap,
-    DenseMap<AMDAIE::BdIdOp, SmallVector<uint32_t>> &bdIdOpToBdIdsMap) {
-  auto tileOp =
-      dyn_cast_if_present<AMDAIE::TileOp>(bdIdOp.getTile().getDefiningOp());
-  if (!tileOp)
-    return bdIdOp.emitOpError()
-           << "doesn't operate on a `amdaie.tile` operation";
-
-  if (!shimTileToGeneratorMap.contains(tileOp.getResult()))
-    return bdIdOp.emitOpError()
-           << "no BD ID generator found for this BD ID op's tile";
-
-  ChannelBdIdGenerator &generator = shimTileToGeneratorMap[tileOp.getResult()];
-  Value value = bdIdOp.getValue();
-  if (auto op = value.getDefiningOp<affine::AffineApplyOp>()) {
-    // If the BD ID is a semi-affine expression.
-    if (bdIdOpToBdIdsMap.contains(bdIdOp)) {
-      for (uint32_t bdId : bdIdOpToBdIdsMap[bdIdOp]) {
-        generator.releaseBdId(bdId);
+    if (!shimTileToGeneratorMap.contains(tileOp.getResult())) {
+      return bdIdOp.emitOpError()
+             << "no BD ID generator found for this BD ID op's tile";
+    }
+
+    ChannelBdIdGenerator &generator =
+        shimTileToGeneratorMap[tileOp.getResult()];
+    Value value = bdIdOp.getValue();
+    if (auto op = value.getDefiningOp<affine::AffineApplyOp>()) {
+      // If the BD ID is a semi-affine expression.
+      if (bdIdOpToBdIdsMap.contains(bdIdOp)) {
+        for (uint32_t bdId : bdIdOpToBdIdsMap[bdIdOp])
+          generator.releaseBdId(bdId);
+      } else {
+        return bdIdOp.emitOpError() << "no BD IDs found for this expression";
       }
     } else {
-      return bdIdOp.emitOpError() << "no BD IDs found for this expression";
+      // Else, must be a constant BD ID.
+      uint32_t bdId = getConstantIndexOrAssert(value);
+      generator.releaseBdId(bdId);
     }
-  } else {
-    // Else, must be a constant BD ID.
-    uint32_t bdId = getConstantIndexOrAssert(value);
-    generator.releaseBdId(bdId);
+    return success();
   }
-  return success();
+
+  /// DmaOps are assigned Bd Ids prior to invoking this function and a map
+  /// `dmaOpToBdIdMap` is maintained that maps a DmaOp to its source/target Bd
+  /// Ids. For each DmaOp in the list `dmaOps`, this API will check the
+  /// `dmaOpToBdIdMap` and release Bd Ids if assigned.
+  LogicalResult releaseAssignedBdIdsInCurrentBatch(
```

Comment:
```suggestion
  LogicalResult releaseAssignedBdIdsInDmaOps(
```
This is not really a batch, either pass a ``DmaBatch` or rename.